src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+        /// sets killsFalseDecision when d being false means other has been proven true
+        /// </summary>
+        private void CheckConsistentDecision(
+            BoundDagDecision d,
+            BoundDagDecision other,
+            out bool permitsTrue,
+            out bool permitsFalse,
+            out bool killsDecisionOnTrueBranch,
+            out bool killsDecisionOnFalseBranch)
         {
-            return other != (object)null && this.Type == other.Type && object.Equals(this.Source, other.Source) && this.Index == other.Index;
+            // innocent until proven guilty
+            permitsTrue = true;
+            permitsFalse = true;
+            killsDecisionOnTrueBranch = false;
+            killsDecisionOnFalseBranch = false;
+
+            // if decisions test unrelated things, there is no implication from one to the other
+            if (d.Input != other.Input)
+            {
+                return;
+            }
+
+            // a test cannot be both true and false
+            if (d == other)
+            {
+                permitsFalse = false;
+                killsDecisionOnTrueBranch = true;
+                return;
+            }
+
+            switch (d)
+            {
+                case BoundNonNullDecision n1:
+                    switch (other)
+                    {
+                        case BoundValueDecision v2:
+                            if (v2.Value == ConstantValue.Null)
+                            {
+                                // once v!=null is false, we know v==null is true and do not need to test it
+                                permitsTrue = false;
+                                killsDecisionOnFalseBranch = true;
+                            }
+                            else
+                            {
+                                // Given that v!=null fails, v==K might succeed
+                                permitsFalse = false;
+                            }
+                            break;
+                        default:
+                            // Once v!=null fails, it must fail a type test
+                            permitsFalse = false;
+                            break;
+                    }
+                    break;
+                case BoundTypeDecision t1:
+                    switch (other)
+                    {
+                        case BoundNonNullDecision n2:
+                            // Once `v is T` is true, v!=null cannot fail
+                            permitsFalse = false;
+                            killsDecisionOnTrueBranch = true;
+                            break;
+                        case BoundTypeDecision t2:
+                            // If T1 could never be T2, then success of T1 implies failure of T2.
+                            var matchPossible = Binder.ExpressionOfTypeMatchesPatternType(Conversions, t1.Type, t2.Type, ref discardedUseSiteDiagnostics, out _);
```

Comment:
>discardedUseSiteDiagnostics [](start = 125, length = 27)

It is not clear why is it safe to discard use-site diagnostics here. It is probably possible to have conversion test fail due to some error that we haven't reported before, therefore causing us to make wrong decision without reporting any error.

Reply:
I'm not sure new use-site diagnostics can arise here. Recorded as an open issue in https://github.com/dotnet/roslyn/issues/23222 .